### PR TITLE
[jcommon] Add capability to pass StatsNameBuilder into JVMStatsExporter

### PR DESCRIPTION
**Pull request by abhimaloo**
[jcommon] Add capability to pass StatsNameBuilder into JVMStatsExporter so that stats from Mbeans can be exported as fb303 with custom naming

JVMStatsExporter exports MBeans into fb303 stats without changing their names. Some of the MBean namings are non-intuitive(weird) like HBase Client uses weird looking "connection ID" into the name. 
This change extends the JVMStatsExporter to receive a StatsNameBuilder object which gets a callback for deciding the name of the stat being exported. This builder can also act as a gatekeeper for stats export i.e if it returns null as name, the stat would not be exported.

### CODE REVIEW

File: stats/src/test/java/com/facebook/stats/mx/TestJVMStatsExporter.java [hunk 1]
```diff
@@ -20,8 +20,10 @@
 import org.testng.Assert;
 import org.testng.annotations.Test;
 
+import javax.management.ObjectName;
```
Comment (twilliamson) [inline stats/src/test/java/com/facebook/stats/mx/TestJVMStatsExporter.java]: Unused import

File: stats/src/test/java/com/facebook/stats/mx/TestJVMStatsExporter.java [hunk 2]
```diff
@@ -62,6 +64,30 @@ public void testFilteredStats() throws Exception {
     Assert.assertTrue(exportedStats.containsKey("jvm.MemoryPool.Code_Cache.UsageThresholdCount"));
     Assert.assertTrue(exportedStats.containsKey("jvm.MemoryPool.Code_Cache.PeakUsage.committed"));
   }
+
+  @Test(groups = "fast")
+  public void testStatNameReplacer() throws Exception {
```
Comment (twilliamson) [inline stats/src/test/java/com/facebook/stats/mx/TestJVMStatsExporter.java]: Can you also add a test that returns `null`?

File: stats/src/test/java/com/facebook/stats/mx/TestJVMStatsExporter.java [hunk 3]
```diff
@@ -62,6 +64,30 @@ public void testFilteredStats() throws Exception {
...
+        for (String attributeName : keys) {
+          name.append('.').append(attributeName);
+        }
+
+        return Optional.of(name.toString());
+      },
+      "java.lang:type=MemoryPool,name=Code Cache"
+    );
+    Map<String, Long> exportedStats = getExportedStats(stats);
+    for (Map.Entry<String, Long> entry : exportedStats.entrySet()) {
```
Comment (twilliamson) [inline stats/src/test/java/com/facebook/stats/mx/TestJVMStatsExporter.java]: Instead of iterating over all the stats, better to assert a specific stat name, e.g.

    Assert.assertNotNull(exportedStats.get("test.Usage.used"), "test.Usage.used");
    Assert.assertNotNull(exportedStats.get("test.Usage.used"), "test.PeakUsage.max");

The current test would pass if no stats were exported.

File: stats/src/main/java/com/facebook/stats/mx/JVMStatsExporter.java [hunk 4]
```diff
@@ -49,6 +54,34 @@
...
+    .add(long.class.getName()).add(Long.class.getName())
+    .add(float.class.getName()).add(Float.class.getName())
+    .add(double.class.getName()).add(Double.class.getName())
+    .build();
+
+  /**
+   * Decides whether to include MBean attributes in the Stats object, and, if
+   * so, what to call them.
+   */
+  public interface StatsNameBuilder {
```
Comment (twilliamson) [inline stats/src/main/java/com/facebook/stats/mx/JVMStatsExporter.java]: Per the [Code Standards](https://github.com/facebook/jcommon/wiki/Coding-Standards#order-of-components-in-a-class), declarations of nested inner classes/interfaces should be after everything else in the class.

File: stats/src/main/java/com/facebook/stats/mx/JVMStatsExporter.java [hunk 5]
```diff
@@ -49,6 +54,34 @@
     .add(SimpleType.DOUBLE)
     .build();
 
+  private static final Set<String> NUMERIC_STRING_TYPES = ImmutableSet.<String>builder()
```
Comment (twilliamson) [inline stats/src/main/java/com/facebook/stats/mx/JVMStatsExporter.java]: What's all this about? The diff is for passing `StatsNameBuilder`, but this seems to be adding some new functionality? Can the `NUMERIC_STRING_TYPES` change be in a separate diff, if we need it?

File: stats/src/main/java/com/facebook/stats/mx/JVMStatsExporter.java [hunk 6]
```diff
@@ -100,12 +171,14 @@ public static Stats createAndBindTo(Stats stats) throws JMException {
...
    * 
    * @param beanNamePattern the bean name pattern used to discover the beans.
+   * @param statsNameBuilder the bean name pattern used to discover the beans.
    * @see javax.management.ObjectName for bean name pattern syntax
    * @see java.lang.management for the list of java platform mbeans
    * 
    * @throws JMException if there are errors querying MBeans or information on them
    */
-  public void exportNumericAttributes(ObjectName beanNamePattern) throws JMException {
+  public void exportNumericAttributes(ObjectName beanNamePattern, StatsNameBuilder statsNameBuilder)
```
Comment (twilliamson) [inline stats/src/main/java/com/facebook/stats/mx/JVMStatsExporter.java]: **This breaks backwards compatibility!** We should **not** change this method signature (see comment above in constructor about how to avoid needing to).

File: stats/src/main/java/com/facebook/stats/mx/JVMStatsExporter.java [hunk 7]
```diff
@@ -71,9 +104,47 @@ public JVMStatsExporter(
     String... beanNamePatterns) 
       throws JMException {
     this.stats = stats;
-    this.statNamePattern = Pattern.compile(statNamePattern);
+    Pattern namePattern = Pattern.compile(statNamePattern);
```
Comment (twilliamson) [inline stats/src/main/java/com/facebook/stats/mx/JVMStatsExporter.java]: Instead of removing the local member variable, just replace it with a `statNameBuilder` member variable. This will allow you to maintain compatibility with the existing API, e.g.:

    private final StatsNameBuilder statsNameBuilder;
    …
    public JVMStatsExporter(
      Stats stats,
      String statNamePattern,
      String... beanNamePatterns
    ) throws JMException {
      this(stats, patternFilter(Pattern.compile(statNamePattern)), beanNamePatterns);
    }
    …
    private static StatsNameBuilder patternFilter(Pattern pattern) {
      return (bean, attribute, key) -> {
        String name = getStatName(bean, attribute, key);
   
        return pattern.matcher(name).matches() ? Optional.of(name) : Optional.empty();
      };
    }

File: stats/src/main/java/com/facebook/stats/mx/JVMStatsExporter.java [hunk 8]
```diff
@@ -49,6 +54,34 @@
...
+     *
+     * @param bean the ObjectName of the bean where the attribute was found
+     * @param attribute the name of the top-level attribute
+     * @param keys if the bean's top-level attribute was {@link CompositeData},
+     *             the keys used to get from there to the value being named.
+     *
+     * @return the name that should be used, or {@link Optional#none()} if it
+     *         should be elided
+     */
+    Optional<String> name(ObjectName bean, String attribute, String... keys);
```
Comment (twilliamson) [inline stats/src/main/java/com/facebook/stats/mx/JVMStatsExporter.java]: The docs are incorrect; this is not parse **all** the keys, but rather called with one key at a time. Think that's the better API anyway, so maybe update to:

    public interface StatsNameBuilder {
      /**
       * Gets the name that should be used in the Stats for an attribute.
       *
       * @param bean the ObjectName of the bean where the attribute was found
       * @param attribute the name of the top-level attribute
       * @param key if the bean's top-level attribute was {@link CompositeData},
       *            the key used to get from there to the value being named.
       *
       * @return the name that should be used, or {@link Optional#none()} if it
       *         should be elided
       */
      Optional<String> name(ObjectName bean, String attribute, String key);
    }

File: stats/src/main/java/com/facebook/stats/mx/JVMStatsExporter.java [hunk 9]
```diff
@@ -180,7 +249,7 @@ private void addStatIfMatches(String statName, Callable<Long> fetcher) {
    * 
    * @return the stat name
    */
-  private static String getStatName(ObjectName beanName, String... attributeNames) {
+  private static String getStatName(ObjectName beanName, List<String> attributeNames) {
```
Comment (twilliamson) [inline stats/src/main/java/com/facebook/stats/mx/JVMStatsExporter.java]: Note: If you do the other comments, you'll probably also want to allow for a `null` attribute name:

    for (String attributeName : attributeNames) {
      if (attributeName != null) {
        builder.append('.').append(attributeName);
      }
    }

File: stats/src/main/java/com/facebook/stats/mx/JVMStatsExporter.java (hunk 5 above)
Comment (abhimaloo) [inline stats/src/main/java/com/facebook/stats/mx/JVMStatsExporter.java]: Messenger code had it, so I though i should add as well. We can delete it for now and see if we need to add it later in another diff
